src/library/components/Modal.tsx

```diff
@@ -0,0 +1,183 @@
+import { cloneElement, ForwardedRef, forwardRef, FunctionComponent, PointerEvent, ReactElement, ReactNode, useCallback, useEffect, useId, useImperativeHandle, useRef, useState } from 'react';
+import Button, { Props as ButtonProps } from './Button';
+import { Props as LinkProps } from './Link';
+
+export interface ModalRef {
+  showModal: () => void;
+  close: () => void;
+  isOpen: boolean;
+}
+
+interface Props {
+  hasClose?: boolean;
+  closeAriaLabel?: string;
+  logo?: string;
+  logoAlt?: string;
+  title: string;
+  positiveAction: ReactElement<ButtonProps | LinkProps>;
+  negativeAction?: ReactElement<ButtonProps | LinkProps>;
+  children: ReactNode;
+  onClose?: () => void;
+  onShowModal?: () => void;
+}
+
+const Modal = forwardRef(({hasClose = true, logo, logoAlt, closeAriaLabel = 'Close modal', title, positiveAction, negativeAction, children, onClose, onShowModal}: Props, ref: ForwardedRef<ModalRef>): JSX.Element => {
+
+  const dialogId = useId();
+
+  const dialogRef = useRef<HTMLDialogElement>(null);
+
+  const [positive, setPositive] = useState<ReactElement<ButtonProps | LinkProps>>();
+  const [negative, setNegative] = useState<ReactElement<ButtonProps | LinkProps>>();
+
+  // Clone positive action to force assertive button style
+  useEffect(() => {
+    if (positiveAction) {
+      const attrs = {
+        variant: 'assertive',
+        className: ''
+      };
+      if ((positiveAction.type as FunctionComponent).displayName === 'Link') {
+        attrs.className = 'bsds-button-assertive'
+      }
+      setPositive(cloneElement(positiveAction, attrs));
+    }
+  }, [positiveAction]);
+
+  // Clone negative action to force default button when close not present and subtle when close present
+  useEffect(() => {
+    if (negativeAction) {
+      const attrs = {
+        variant: hasClose ? 'subtle' : 'default',
+        className: ''
+      };
+      if ((negativeAction.type as FunctionComponent).displayName === 'Link') {
+        attrs.className = hasClose ? 'bsds-button-subtle' : 'bsds-button'
+      }
+      setNegative(cloneElement(negativeAction, attrs));
+    }
+  }, [negativeAction, hasClose]);
+
+  /**
+   * Opens or closes the modal, and manages overflowY on body.
+   * When modal is open overflowY is hidden, when modal is closed overflowY is auto.
+   * @param {boolean} shouldShow - Modal opens when true, closes when false.
+   */
+  const showDialog = useCallback((shouldShow: boolean) => {
+    const body = document.querySelector('body');
+    if (body) {
+      if (shouldShow) {
+        dialogRef.current?.showModal();
+        body.style.overflowY = 'hidden';
```

Comment:
We could consider adding and applying utility classes instead of using the style attr. Let's keep an eye out for other places we might do this.